### PR TITLE
Cache output of assembly

### PR DESCRIPTION
**Pull request by Duhemm**
Previously, the build would rebuild over and over the uber jar with the
shaded protobuf3 distribution. With this patch, this is no longer the
case, as the output is cached locally.

------

The issue with assembly re-creating the same uber-jar on (nearly) every run was discovered using the Develocity Timeline feature, which displays task executions on different swimlanes.

Before this fix, the assembly task is clearly shown as being a bottleneck in the build process:
Screenshot 2024-02-02 at 16 19 38

After this patch is applied, the bottleneck is gone:
Screenshot 2024-02-02 at 16 19 51

After the bottleneck is gone, common tasks (eg. `test`) start faster: Gathering all tests’ names goes from 3.7s to 3.2s (-15%).

### CODE REVIEW

**Comment by pjfanning** [comment]
relates to #1008

File: project/Protobuf.scala
Comment (Roiocam) [inline project/Protobuf.scala:41]: This line would be execute osgi bundle unexpected?

**Review by Roiocam — COMMENTED** [review]
Looks great, Thanks.

**Comment by mdedetrich** [comment]
Thanks for the PR, @Roiocam fyi this is precisely what I was talking about earlier wrt sbt's inbuilt caching functionality

File: project/Protobuf.scala
Comment (mdedetrich) [reply project/Protobuf.scala:41]: This to me also looks suspicious, @Duhemm whats the context behind this change?

Comment (Duhemm) [reply project/Protobuf.scala:41]: `Compile / packagebin` is cached, whereas `assembly` is not. Both tasks produce the same output:

On main:
```
pekko > show protobuf-v3 / Compile / assembly
(...)
    [info] /Users/martin/Documents/Projects.nosync/apache/incubator-pekko/protobuf-v3/target/scala-2.13/stripped/pekko-protobuf-v3-assembly-1.1.0-M0+329-231b9a99-SNAPSHOT.jar
$ md5sum < /Users/martin/Documents/Projects.nosync/apache/incubator-pekko/protobuf-v3/target/scala-2.13/stripped/pekko-protobuf-v3-assembly-1.1.0-M0+329-231b9a99-SNAPSHOT.jar
87fdacb73c114cf95d088cb31c164622  -

pekko > show protobuf-v3 / Compile / packageBin
(...)
    [info] /Users/martin/Documents/Projects.nosync/apache/incubator-pekko/protobuf-v3/target/scala-2.13/stripped/stripped/pekko-protobuf-v3-assembly-1.1.0-M0+329-231b9a99-SNAPSHOT.jar
$ md5sum < /Users/martin/Documents/Projects.nosync/apache/incubator-pekko/protobuf-v3/target/scala-2.13/stripped/stripped/pekko-protobuf-v3-assembly-1.1.0-M0+329-231b9a99-SNAPSHOT.jar
87fdacb73c114cf95d088cb31c164622  -
```

With this pull request:
```
pekko > show protobuf-v3 / Compile / assembly
(...)
[info] /Users/martin/Documents/Projects.nosync/apache/incubator-pekko/protobuf-v3/target/scala-2.13/stripped/pekko-protobuf-v3-assembly-1.1.0-M0+330-f4e48f91-SNAPSHOT.jar
$ md5sum < /Users/martin/Documents/Projects.nosync/apache/incubator-pekko/protobuf-v3/target/scala-2.13/stripped/pekko-protobuf-v3-assembly-1.1.0-M0+330-f4e48f91-SNAPSHOT.jar
aec27c8ab1a2dd13fa5b2f15044089d1  -

pekko > show protobuf-v3 / Compile / packageBin
(...)
[info] /Users/martin/Documents/Projects.nosync/apache/incubator-pekko/protobuf-v3/target/scala-2.13/stripped/stripped/pekko-protobuf-v3-assembly-1.1.0-M0+330-f4e48f91-SNAPSHOT.jar
$ md5sum < /Users/martin/Documents/Projects.nosync/apache/incubator-pekko/protobuf-v3/target/scala-2.13/stripped/stripped/pekko-protobuf-v3-assembly-1.1.0-M0+330-f4e48f91-SNAPSHOT.jar
aec27c8ab1a2dd13fa5b2f15044089d1  -
```

So depending on `packageBin` rather than `assembly` should produce the same result without requiring to re-build the jar over and over again.

What's your concern with this change?

Comment (mdedetrich) [reply project/Protobuf.scala:41]: > What's your concern with this change?

Its just that this area of pekko has historically been quite brittle largely due to how sbt + sbt-osgi is designed (see https://github.com/sbt/sbt-osgi/issues/104).

I am quite sure that this is a non issue, its just we have been burned in the past with subtle breakages but @Roiocam has added regression tests so I will approve the PR.

**Review by mdedetrich — APPROVED** [review]
lgtm but lets get approval from @Roiocam before merging to double check nothing is amiss.

File: project/Protobuf.scala
Comment (Roiocam) [reply project/Protobuf.scala:41]: protobufV3 uses a specific osgi packaging approache instead of relying on packageBin, so I have no problem with this PR.

**Review by Roiocam — APPROVED** [review]
LGTM, Thanks.

**Comment by mdedetrich** [comment]
Thanks for the PR then, merging!

**Pull request merged**
